File: Robotics_Kinematics_and_Dynamics_Course/Chapter11/Lesson1/Chapter11_Lesson1.m
function qddot = pendulumEoM(t, x, tau, m, l, g)
% x = [q; qdot]
q    = x(1);
qdot = x(2);

M = m * l^2;
G = m * g * l * sin(q);
Cqdot = 0;   % no Coriolis term for 1-DOF

qddot = (tau - Cqdot - G) / M;
qddot = [qdot; qddot];
end

% Example numerical simulation with ODE45
m = 1.0; l = 1.0; g = 9.81;
tau_const = 0;

f = @(t, x) pendulumEoM(t, x, tau_const, m, l, g);
[t_sol, x_sol] = ode45(f, [0 10], [0.5; 0.0]);

plot(t_sol, x_sol(:,1)); xlabel('t'); ylabel('q(t)');
title('Pendulum angle using Lagrange-Euler EoM');
      

File: Robotics_Kinematics_and_Dynamics_Course/Chapter11/Lesson1/Chapter11_Lesson1.java
public class PendulumDynamics {

    public static class Params {
        public double m;
        public double l;
        public double g;

        public Params(double m, double l, double g) {
            this.m = m;
            this.l = l;
            this.g = g;
        }
    }

    // qddot = (tau - m g l sin(q)) / (m l^2)
    public static double accel(double q, double qdot, double tau, Params p) {
        double M = p.m * p.l * p.l;
        double G = p.m * p.g * p.l * Math.sin(q);
        double Cqdot = 0.0;
        return (tau - Cqdot - G) / M;
    }

    public static void main(String[] args) {
        Params p = new Params(1.0, 1.0, 9.81);
        double q = 0.5;
        double qdot = 0.0;
        double tau = 0.0;
        double qddot = accel(q, qdot, tau, p);
        System.out.println("qddot = " + qddot);
    }
}
      
